src/projection.jl

```diff
@@ -178,14 +189,21 @@ end
 # no structure worth re-imposing. Then any array is acceptable as a gradient.
 
 # For arrays of numbers, just store one projector:
-function ProjectTo(x::AbstractArray{T}) where {T<:Number}
-    element = T <: Irrational ? ProjectTo{Real}() : ProjectTo(zero(T))
-    if element isa ProjectTo{<:AbstractZero}
-        return ProjectTo{NoTangent}() # short-circuit if all elements project to zero
+function ProjectTo(x::AbstractArray{T,N}) where {T<:Number,N}
+    element = _eltype_projectto(T)
+    S = project_type(element)  # new idea -- for any number, S is enough. 
+                               # Store .element for now too, although it's redundant? Reconstruct from eltype?
+    if axes(x) isa NTuple{N,Base.OneTo{Int}}
+        return ProjectTo{AbstractArray{S,N}}(; element=element, axes=axes(x))
```

Comment:
Here `S` and `p.element` are the same information: for every `S<:Number`, the projector can be reconstructed as `ProjectTo(zero(S))`. So perhaps tidier not to store it.

We can alter `getproperty` to do this. In which case, perhaps the projector for an array of arrays (which has an array `p.elements` instead) should also define it, via `mapreduce(project_type, promote_type, projectors)`, even if that ends up trivial? 

If you can always call `p.elements`, then you can insert that into some in-place rules, `dx -> dx .+= ρ.(dy .* conj.(stuff...))`.